pkg/controller/service/service_controller.go

```diff
@@ -257,8 +257,6 @@ func (s *ServiceController) processServiceUpdate(cachedService *cachedService, s
 		}
 	}
 
-	// cache the service, we need the info for service deletion
```

Comment:
Good catch for the no-op retry. Though I'm a bit concerned about this comment. What happen if:
1. `createLoadBalancerIfNeeded()` fails hence `cachedService.state` is nil.
2. Before a successful retry, service is deleted on kube-apiserver.
3. `processLoadBalancerDelete()` is triggered and it references `cachedService.state` for deletion. This may either:
     - Panic on nil pointer.
     - Fail to cleanup the half-created LB resources as we have no information about this service anymore: https://github.com/kubernetes/kubernetes/blob/ccb15fb498352cad3f582cce0ade4e99efe0f82f/pkg/controller/service/service_controller.go#L798-L816

Reply:
> Fail to cleanup the half-created LB resources as we have no information about this service anymore

If there's no reconcile loop on the LB resources that compares against actually present services, this seems like an existing gap, since the service controller can restart, miss deletion events, and leave orphaned LBs

Reply:
Agree that the service controller restart case is an existing gap, but the half-created LB case may not be. A common failure scenario I could think of is:
1. User has enough firewall quota in project but runs out of LB quota.
2. A LoadBalancer type service is created.
3. `createLoadBalancerIfNeeded()` created the corresponding firewall, but hit quota exceeded error while provisioning LB.
4. User waits for LB service to come up for a while but it doesn't, so he/she deletes the service.
5. Firewall or any other created LB resources are leaked.

Which sounds like a bad user experience.

Reply:
According to this comment 
> 	// NOTE: Since we update the cached service if and only if we successfully
	// processed it, a cached service being nil implies that it hasn't yet
	// been successfully processed.
the PR changes are valid; however, as @MrHohn pointed out the controller isn't actually setup to handle that condition. The cached state is only ever nil on the initial processing of a service after a cold cache or after creation. 

I wish I had finished the finalizer PR https://github.com/kubernetes/kubernetes/pull/54569 for 1.9 since it would have made all these valid cleanup concerns a non-issue. As part of that I'm able to remove the dependency on the cached object for deletion and cleanup other logic. 

I can't see any easy way to solve the current problem until that PR is finished especially since we're already in code freeze. Unless anyone has any ideas I think the best solution is to revert #55336, which removes the no-op case but keeps the race condition fixed in that PR. That condition seems way less likely than the scenario @MrHohn pointed out above so it's the "lesser of the two evils" and we can include all the fixes in 1.10. Thoughts?